### PR TITLE
Add a --debug argument and some logging when retrieving CI status

### PR DESCRIPTION
**Pull request by pypingou**

### CODE REVIEW

File: bodhi/client/bindings.py [hunk 1]
```diff
@@ -172,6 +174,12 @@ def __init__(self, base_url=BASE_URL, username=None, password=None, staging=Fals
         if base_url[-1] != '/':
             base_url = base_url + '/'
 
+        if debug:
+            ch = logging.StreamHandler()
+            ch.setLevel(logging.DEBUG)
+            log.addHandler(ch)
+            log.setLevel(logging.DEBUG)
```
Comment (bowlofeggs) [inline bodhi/client/bindings.py]: libraries should not set up logging handlers or set log levels. This library is not just used by the bodhi cli and so it should be unopinionated about logging. It's OK for the lib to log things at debug level, but the logger should not be configured by it.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]

File: bodhi/client/bindings.py (hunk 1 above)
Comment (pypingou) [reply bodhi/client/bindings.py]: Any preference where we could configure it?
I thought this was a good place since it was common with all the actions performed by the CLI, but I'm fine w/ changing this :)

**Comment by bowlofeggs** [comment]
On 07/25/2018 11:03 AM, Pierre-Yves Chibon wrote:
> Any preference where we could configure it?

We should do it in bodhi/client/__init__.py. One way I can think of to
do it as a common action so you don't have to repeat code is you could
set up a validator on the --debug flag that does this if --debug is set.
There's already a validator in there on at least one flag as an example.

**Comment by pypingou** [comment]
> One way I can think of to do it as a common action so you don't have to repeat code is you could set up a validator on the --debug flag that does this if --debug is set. There's already a validator in there on at least one flag as an example.

Is the validator you have in mind the ``callback`` used for the ``--url`` argument? If not then I need to look further for this validator as I think I didn't find it then :s

I took a little bit of a different approach in the mean while, I've added the ``--debug`` argument to the ``cli()`` method, so all actions have it, but it needs to be specified at the beginning of the argument list, for example:
``bodhi --debug updates waive  python-arrow-0.8.0-5.fc28 --show --staging``
works while:
``bodhi updates waive  python-arrow-0.8.0-5.fc28 --show --staging --debug`` does not.

Again, if you prefer the validator approach I'll look into it some more :)

**Comment by bowlofeggs** [comment]
On 07/26/2018 07:23 AM, Pierre-Yves Chibon wrote:
> Is the validator you have in mind the |callback| used for the |--url|
> argument? If not then I need to look further for this validator as I
> think I didn't find it then :s

Yeah it's the callback system.

> I took a little bit of a different approach in the mean while, I've
> added the |--debug| argument to the |cli()| method, so all actions have
> it, but it needs to be specified at the beginning of the argument list,
> for example:
> |bodhi --debug updates waive python-arrow-0.8.0-5.fc28 --show --staging|
> works while:
> |bodhi updates waive python-arrow-0.8.0-5.fc28 --show --staging --debug|
> does not.
> 
> Again, if you prefer the validator approach I'll look into it some more :)

I think I prefer the validator approach since it will be more consistent
with how the rest of Bodhi's flags work, unless you can find a way to
make it work both ways (I don't know a way to do it off the top of my
head, but it might be possible.)

**Comment by pypingou** [comment]
I've pushed a commit porting ``--debug`` to use a callback mechanism as ``--url`` does. You will see it requires more changes than the ``cli()`` approach. If you're fine with that approach, I'll merge the third commit into the first one to clean up the history :)

**Comment by pypingou** [comment]
Rebased to fix the conflict

**Comment by pypingou** [comment]
--hm these tests are definitively passing locally :(--

Nevermind the last comment from jenkins confused me, they are also passing on jenkins now it seems :)

File: bodhi/client/__init__.py [hunk 2]
```diff
@@ -453,7 +488,9 @@ def query(url, mine=False, rows=None, **kwargs):
         mine (Boolean): If the --mine flag was set
         kwargs (dict): Other keyword arguments passed to us by click.
     """
+    kwargs.pop('debug')
     client = bindings.BodhiClient(base_url=url, staging=kwargs['staging'])
+
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: This is an unrelated style change, so let's leave it as it is.

File: bodhi/client/__init__.py [hunk 3]
```diff
@@ -295,7 +327,7 @@ def new(user, password, url, **kwargs):
                        True.
         kwargs (dict): Other keyword arguments passed to us by click.
     """
-
+    kwargs.pop('debug')
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: Is this what you meant by being a larger change? If so, a cleaner way to do this is to add ```debug``` as a parameter to each of these functions so it doesn't end up in ```kwargs```.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]

File: bodhi/client/__init__.py (hunk 3 above)
Comment (pypingou) [reply bodhi/client/__init__.py]: Good idea, will fix

File: bodhi/client/__init__.py (hunk 2 above)
Comment (pypingou) [reply bodhi/client/__init__.py]: ok

File: bodhi/client/__init__.py [hunk 4]
```diff
@@ -244,7 +274,8 @@ def composes():
 @staging_option
 @click.option('-v', '--verbose', is_flag=True, default=False, help='Display more information.')
 @url_option
-def list_composes(url, staging, verbose):
+@debug_option
+def list_composes(url, staging, verbose, debug):
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py:279]: Note that we still document parameters in the "developer docs" section below, here and elsewhere in the file.

File: bodhi/client/__init__.py [hunk 5]
```diff
@@ -295,7 +327,6 @@ def new(user, password, url, **kwargs):
                        True.
         kwargs (dict): Other keyword arguments passed to us by click.
     """
-
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: This is an unrelated style change.

File: bodhi/client/__init__.py [hunk 6]
```diff
@@ -432,9 +464,10 @@ def edit(user, password, url, **kwargs):
 @click.option('--mine', is_flag=True, help='Show only your updates')
 @staging_option
 @url_option
+@debug_option
 @add_options(pagination_options)
 @handle_errors
-def query(url, mine=False, rows=None, **kwargs):
+def query(url, mine=False, rows=None, debug=False, **kwargs):
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: Let's not define the default for ```debug``` in two places. If you move this to be after ```url```, you don't need to give it a default.

File: bodhi/tests/client/test___init__.py [hunk 7]
```diff
@@ -324,6 +324,46 @@ def test_severity_flag(self, send_request):
         ]
         self.assertEqual(send_request.mock_calls, calls)
 
+    @mock.patch.dict(client_test_data.EXAMPLE_UPDATE_MUNCH, {'severity': 'urgent'})
+    @mock.patch.dict(os.environ, {'BODHI_URL': 'http://example.com/tests/'})
+    @mock.patch('bodhi.client.bindings.BodhiClient.csrf',
+                mock.MagicMock(return_value='a_csrf_token'))
+    @mock.patch('bodhi.client.bindings.BodhiClient.send_request',
+                return_value=client_test_data.EXAMPLE_UPDATE_MUNCH, autospec=True)
+    def test_severity_and_debug_flags(self, send_request):
```
Comment (bowlofeggs) [inline bodhi/tests/client/test___init__.py]: This test is really about testing the debug flag, so let's name it ```test_debug_flag``` for simplicity.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
Just a few more minor changes.

File: bodhi/client/__init__.py (hunk 4 above)
Comment (pypingou) [reply bodhi/client/__init__.py:279]: Will add

File: bodhi/client/__init__.py (hunk 5 above)
Comment (pypingou) [reply bodhi/client/__init__.py]: will remove

File: bodhi/client/__init__.py (hunk 6 above)
Comment (pypingou) [reply bodhi/client/__init__.py]: Will fix

File: bodhi/tests/client/test___init__.py (hunk 7 above)
Comment (pypingou) [reply bodhi/tests/client/test___init__.py]: Will adjust

**Review by pypingou — COMMENTED** [review]
thanks for the review

**Comment by pypingou** [comment]
I think I got all of your comments, I also merged the new approach into the old one making the history cleaner.

**Comment by bowlofeggs** [comment]
I forgot to mention when reviewing this one - can you also squash your
commits into 1? Thanks!

File: bodhi/client/__init__.py [hunk 8]
```diff
@@ -53,13 +57,37 @@ def _warn_if_url_and_staging_set(ctx, param, value):
...
+    """
+    Set up the logging level to "debug".
+
+    This allows us to print more information on the user's screen and thus helps following what is
+    going on.
+
+    Args:
+        ctx (click.core.Context): The Click context. Unused.
+        param (click.core.Option): The option being handled. Unused.
+        value (unicode): The value of the --debug flag.
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: This is a ```bool```, not a ```unicode```.

File: bodhi/client/__init__.py [hunk 9]
```diff
@@ -53,13 +57,37 @@ def _warn_if_url_and_staging_set(ctx, param, value):
...
+    Returns:
+        unicode: The value of the --debug flag.
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: This is also a ```bool```.

File: bodhi/client/__init__.py [hunk 10]
```diff
@@ -53,13 +57,37 @@ def _warn_if_url_and_staging_set(ctx, param, value):
...
+    return value
+
+
 url_option = click.option('--url', envvar='BODHI_URL', default=bindings.BASE_URL,
                           help=('URL of a Bodhi server. Ignored if --staging is set. Can be set '
                                 'with BODHI_URL environment variable'),
                           callback=_warn_if_url_and_staging_set)
 staging_option = click.option('--staging', help='Use the staging bodhi instance',
                               is_flag=True, default=False)
-
```
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: Unrelated style change.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
Just 3 small things I'd like to fix, then we'll be good to merge.

**Comment by bowlofeggs** [comment]
I will fix these three things myself.

**Review by bowlofeggs — APPROVED** [review]

**Comment by pypingou** [comment]
Thanks @bowlofeggs :)

**Comment by bowlofeggs** [comment]
These patches are planned to be included in the upcoming 3.10.0 release: #2556.

**Comment by bowlofeggs** [comment]
These patches are available in Copr:

https://copr.fedorainfracloud.org/coprs/bowlofeggs/bodhi-pre-release/
